drv/lpc55-update-server/src/images.rs

```diff
@@ -26,79 +47,617 @@ extern "C" {
+                        .checked_add(remainder as u32)
+                        .ok_or(UpdateError::OutOfBounds)?;
+                    if span.stored.contains(&start)
+                        && (span.stored.start..=span.stored.end).contains(&end)
+                    {
+                        indirect_flash_read(flash, start, buffer)?;
+                    } else {
+                        return Err(UpdateError::OutOfBounds);
+                    }
+                }
+                Ok(())
+            }
+        }
+    }
+
+    /// Get the rounded up length of an LPC55 image if present.
+    pub fn padded_image_len(&self) -> Result<u32, UpdateError> {
+        let vectors = match self.accessor {
+            Accessor::Flash { .. } => {
+                let buffer =
+                    &mut [0u8; core::mem::size_of::<ImageVectorsLpc55>()];
+                self.read_bytes(0u32, buffer.as_bytes_mut())?;
+                ImageVectorsLpc55::read_from_prefix(&buffer[..])
+                    .ok_or(UpdateError::OutOfBounds)
+            }
+            Accessor::Ram { buffer, .. } | Accessor::Hybrid { buffer, .. } => {
+                ImageVectorsLpc55::read_from_prefix(buffer)
+                    .ok_or(UpdateError::OutOfBounds)
+            }
+        }?;
+        let len = vectors.image_length().ok_or(UpdateError::BadLength)?;
+        round_up_to_flash_page(len).ok_or(UpdateError::BadLength)
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub struct Epoch {
+    value: u32,
+}
+
+/// Convert from the ImageHeader.epoch format
+impl From<u32> for Epoch {
+    fn from(number: u32) -> Self {
+        Epoch { value: number }
+    }
+}
+
+/// Convert from the caboose EPOC value format.
+//
+// Invalid EPOC values converted to Epoch{value:0} include:
+//   - empty slice
+//   - any non-ASCII-digits in slice
+//   - any non-UTF8 in slice
+//   - leading '+' normally allowed by parse()
+//   - values greater than u32::MAX
+//
+//   Hand coding reduces size by about 950 bytes.
+impl From<&[u8]> for Epoch {
+    fn from(chars: &[u8]) -> Self {
+        let epoch =
+            if chars.first().map(|c| c.is_ascii_digit()).unwrap_or(false) {
+                if let Ok(chars) = core::str::from_utf8(chars) {
+                    chars.parse::<u32>().unwrap_or(0)
+                } else {
+                    0
+                }
+            } else {
+                0
+            };
+        Epoch::from(epoch)
+    }
+}
+
+impl Epoch {
+    pub fn can_update_to(&self, next: Epoch) -> bool {
+        self.value <= next.value
+    }
+}
+
+/// Check a next image against an active image to determine
+/// if rollback policy allows the next image.
+/// If ImageHeader and Caboose are absent or Caboose does
+/// not have an `EPOC` tag, then the Epoch is defaulted to zero.
+/// This test is also used when the header block first arrives
+/// so that images can be rejected early, i.e. before any flash has
+/// been altered.
+pub fn check_rollback_policy(
```

Comment:
It seems like this function is only ever useful when `complete` is `true`; otherwise, we can't read the epoch tag from the new image's caboose (so we have to keep loading more of the image).  Could we remove the `complete` flag and only call the function on complete images?